src/main/kotlin/uk/gov/justice/digital/hmpps/sentenceplan/services/GoalService.kt

```diff
@@ -190,55 +190,87 @@ class GoalService(
 
   @Transactional
   fun updateGoalStatus(goalUuid: UUID, updatedGoal: Goal): GoalEntity {
-    // 1. if we have a note value, create a new note of the correct type
-    // 2. update the current goal status
+    if (updatedGoal.status == null) {
+      throw ValidationException("Updated Goal status must not be null")
+    }
 
-    if (updatedGoal.status == null || (updatedGoal.status == GoalStatus.REMOVED && updatedGoal.note.isNullOrEmpty())) {
-      throw ValidationException("One or more required fields are null")
+    val goalEntity = goalRepository.getGoalByUuid(goalUuid)
+
+    goalEntity.notes.add(
+      GoalNoteEntity(note = updatedGoal.note!!, goal = goalEntity, type = GoalNoteType.PROGRESS),
+    )
+    goalEntity.status = updatedGoal.status!!
+    goalEntity.statusDate = LocalDateTime.now()
+
+    return goalRepository.save(goalEntity)
+  }
+
+  @Transactional
+  fun achieveGoal(goalUuid: UUID, updatedGoal: Goal): GoalEntity {
+    if (updatedGoal.status != GoalStatus.ACHIEVED) {
```

Comment:
I wonder if we even need this check. If we weren't handling the legacy of updateGoalStatus then this method should only really need the goalUuid and the note text, given this method is being explicitly called. Then we could set goalEntity.status specifically to ACHIEVED.

Reply:
Resolved in [1e4b4d7](https://github.com/ministryofjustice/hmpps-sentence-plan/pull/199/commits/1e4b4d76850be0d3b55bc82a5a91be5f74093f2c)